### ISSUE TITLE
proposal: do not pack SIMD vectors

### ISSUE THREAD

**Issue by Snektron** [ziglang/zig#18652]
Currently, `@Vector(N, T)` is packed together. For example:
```zig
export fn square() u32 {
    return @bitSizeOf(@Vector(11, u3));
}
```
[returns](https://godbolt.org/z/ov8fvWM9c) 33. This seems counter intuitive to me for several reasons:
- `[N]T` is not packed either.
- I see no good reason for vectors to be packed.
- The performance impact is highly questionable. For instance, consider
    ```zig
    export fn square(num: u32) u32 {
        return @reduce(.Add, @as(@Vector(11, u3), @bitCast(@as(u33, num))));
    }
    ```
    This [generates](https://godbolt.org/z/z3KqvGabP) an excessive amount of bit shifts
 
In my opinion, vectors should essentially be a bag of scalars that you want to perform the same operation on, and not provide any layout guarantees at all. This would enable compilers to lower `@Vector(11, u3)` to `@Vector(11, u8)`, and omit these expensive shifts (and a whole lot of headaches).

An important edge case here are `@Vector(N, bool)` and `@Vector(N, u1)`. I suspect the reason why the above are packed at all, is to provide the guarantee that those are backed by an integer (and that all operations on it are bitwise). This makes sense to me, and I don't think that we should remove that. I see three main paths forward:
- Make only `@Vector(N, bool)` be guaranteed to be backed by `uN`, and allow `@bitcast`ing between these.
- Make only `@Vector(N, bool)` and `@Vector(N, u1)` be backed by `uN`, and allow `@bitcast`ing between these.
- Add no such guarantee at all (leave it up to the backend), and provide utilities (functions or built-ins) to between `uN` and vectors.

In all cases, I think we only need to remove the capability to `bitcast` between vectors and integers.

**Comment by andrewrk** [comment]
`@bitSizeOf` reports how many bits of logical information there are. For example, `i9` returns 9 for bit size of, while it actually takes up 2 bytes of actual memory (16 bits) in its representation. I can still make sense of your initial point however if we look at `@sizeOf` instead and make the same observation.

> In my opinion, vectors should essentially be a bag of scalars that you want to perform the same operation on, and not provide any layout guarantees at all. This would enable compilers to lower `@Vector(11, u3)` to `@Vector(11, u8)`, and omit these expensive shifts (and a whole lot of headaches).

This is status quo.

> Add no such guarantee at all (leave it up to the backend), and provide utilities (functions or built-ins) to between uN and vectors.

Again this is status quo.

> In all cases, I think we only need to remove the capability to bitcast between vectors and integers.

But that's the utility to convert between uN and vectors that you mentioned above.

**Comment by Snektron** [comment]
Alright, I guess I misunderstood the meaning of bitcast in this context. This doesn't change the fact that LLVM still reasons about it in a packed way, which I still believe should be corrected.

**Comment by andrewrk** [comment]
I agree, that should be corrected. Happy to make any clarifications to the lang ref or type up some spec text if you need it to get unblocked. Let me know how I can help. But it sounds like everything you want to do in the backend is already legal.

**Comment by mlugg** [comment]
Status quo vectors are a little bit of a mess - I discussed this with @andrewrk and @jacobly0 in a compiler meeting a while back. Some parts of the language (bitcasting) assume they are packed, and some (vector index in bit-pointers) do not.

If I remember correctly, Jacob proposed behavior something like this:
* `@Vector(n, bool)` has a well-defined layout and bit-packs; this is fine on any realistic platform, and can be useful for `@bitCast` shenanigans
* Any other vector has undefined layout; the backend can choose how best to lower it. This probably usually means they should be aligned up to the next power-of-two integer. (Since they don't have well-defined layout, `@bitCast`ing to or from one is a compile error, and reinterpreting one through a pointer is UB.)

The only compiler change that would be needed here to match the language spec is to change `Type.hasWellDefinedLayout`, which currently just returns `true` for vectors.

Another problematic thing with vectors is the "vector index" field on bit-pointers. I don't love it in general - I think it would make more sense to return bit-pointers, even if per the above rules this makes the exact type returned backend-dependent (related: #16271) - but the bigger problem is that the vector index is permitted to be runtime-known (represented by writing it as a `?`), which completely breaks how bit-pointers are supposed to work, and can cause completely incorrect behavior even today if you do `&unaligned_vector[runtime_index]`. (Fixing this basically *would* be #16271.)

**Comment by p7r0x7** [comment]
I've been thinking about this a lot, as my upcoming VP9 encoder project heavily relies on SIMD and @Vector(). I would love to use arbitrary-bit-wide integers efficiently (because the compiler "reduces"/"promotes" them to 2^x bit sizes).

**Comment by alexrp** [comment]
While auditing `volatile` semantics in the LLVM backend, I noticed that we're in violation here:

https://github.com/ziglang/zig/blob/ab89af3d397291f0126ad17d3787a51284457884/src/codegen/llvm.zig#L8398-L8419

Note how an operation that looks like it should be a single store in code (`vec_ptr[i] = x`) turns into a load + store. We can fix this for some types of vectors, but not in the general case due to vectors being packed.

So I think I'm in favor of some form of this proposal.

**Comment by mlugg** [comment]
The accepted semantics for `@bitSizeOf`, `@bitCast`, etc, are to operate on the "logical bit layout" of a value (see #19755), so the value it returns does not reflect how it is represented in-memory. Right now I think the LLVM backend *does* bit-pack vectors, but (once that proposal is fully implemented) it need not do so. Changing that can be another issue (toolchain enhancement, not a proposal) on Codeberg if anyone wants to open it.
